src/animation.rs

```diff
@@ -0,0 +1,532 @@
+
+    /// Targets the orientation property of an [`Object`].
+    ///
+    /// The corresponding keyframe values must be [`Quaternion`] or [`Euler`].
+    ///
+    /// [`Object`]: ../object/struct.Object.html
+    /// [`Quaternion`]: enum.Values.html#variant.Quaternion
+    /// [`Euler`]: enum.Values.html#variant.Euler
+    Orientation,
+
+    /// Targets the scale property of an [`Object`].
+    ///
+    /// The corresponding keyframe values must be [`Scalar`].
+    ///
+    /// [`Object`]: ../object/struct.Object.html
+    /// [`Scalar`]: enum.Values.html#variant.Scalar
+    Scale,
+}
+
+/// An index into the frames of a track.
+enum FrameRef {
+    /// The time is before the start of the frames.
+    Unstarted,
+
+    /// The time corresponds to the given frame index.
+    InProgress(usize),
+
+    /// The time is after the end of the last frame.
+    Ended,
+}
+
+/// The keyframe values of a [`Track`].
+///
+/// [`Track`]: struct.Track.html
+#[derive(Clone, Debug)]
+pub enum Values {
+    /// Euler angle keyframes in radians.
+    Euler(Vec<mint::EulerAngles<f32, IntraXyz>>),
+
+    /// Quaternion keyframes.
+    Quaternion(Vec<mint::Quaternion<f32>>),
+
+    /// Scalar keyframes.
+    ///
+    /// ## Note
+    ///
+    /// Only uniform scaling is supported, hence the glTF importer takes the Y axis as the scaling
+    /// direction, ignoring any scaling in the X and Z axes.
+    Scalar(Vec<f32>),
+
+    /// 3D vector keyframes.
+    Vector3(Vec<mint::Vector3<f32>>),
+}
+
+/// Message data sent from `Action` to `Mixer` over a channel.
+enum Operation {
+    Enable,
+    Disable,
+    Pause,
+    Play,
+    SetLoopMode(LoopMode),
+}
+
+/// Controls the playback properties of an animation
+pub struct Action {
+    /// Message channel to hub.
+    tx: mpsc::Sender<Message>,
+
+    /// Pointer to parent `Mixer` struct.
+    pointer: froggy::Pointer<ActionData>,
+}
+
+/// Internal data for an animation action.
+struct ActionData {
+    /// The animation data for this action.
+    pub clip: Clip,
+
+    /// Specifies whether the action is enabled or disabled.
+    ///
+    /// A disabled action has no impact.
+    pub enabled: bool,
+
+    /// Specifies the looping behaviour of this action.
+    pub loop_mode: LoopMode,
+
+    /// Specifies whether the action is paused.
+    pub paused: bool,
+
+    /// The local time of this action in seconds, starting at 0.0.
+    ///
+    /// This value get clamped or wrapper to [0.0, clip.duration] depending on
+    /// the loop mode.
+    pub local_time: f32,
+
+    /// Time scaling factor.
+    pub local_time_scale: f32,
+
+    // Unimplemented properties
+    // ------------------------
+    // * weight
+    // * zero_slope_at_end
+    // * zero_slope_at_start
+}
+
+/// A reusable set of keyframe tracks which represent an animation.
+#[derive(Clone, Debug)]
+pub struct Clip {
+    /// A name for this clip.
+    pub name: Option<String>,
+
+    /// The animation keyframe tracks.
+    pub tracks: Vec<Track>,
```

Comment:
can we have `pub tracks: Vec<(Track, Object)>` so that object mutability is moved out of the track data?

Reply:
Sure.